drivers: sensor: ti: ina230: Add support for INA236

jurenat opened the pull request:
This commit adds support for INA236 into the existing INA230 driver. These two chips are similar enough to share most of the code. The device can be defined the same way as INA230 and we only have the extra option to select the high-precision mode.

```
ina236: ina236@40 {
		status = "okay";
		compatible = "ti,ina236", "ti,ina230";
		reg = <0x40>;
		adc-mode = "Bus and shunt voltage continuous";
		vbus-conversion-time-us = <1100>;
		vshunt-conversion-time-us = <1100>;
		avg-count = <1>;
		current-lsb-microamps = <1000>;
		rshunt-micro-ohms = <15000>;
		alert-gpios = <&gpiod 0 GPIO_ACTIVE_LOW>;
		high-precision;
	};
```

jurenat committed:
drivers: sensor: ti: ina230: Add support for INA236

This commit adds support for INA236 into the existing INA230 driver.
These two chips are similar enough to share most of the code.
The device can be defined the same way as INA230 and we only have
the extra option to select the high-precision mode.

```
ina236: ina236@40 {
		status = "okay";
		compatible = "ti,ina236", "ti,ina230";
		reg = <0x40>;
		adc-mode = "Bus and shunt voltage continuous";
		vbus-conversion-time-us = <1100>;
		vshunt-conversion-time-us = <1100>;
		avg-count = <1>;
		current-lsb-microamps = <1000>;
		rshunt-micro-ohms = <15000>;
		alert-gpios = <&gpiod 0 GPIO_ACTIVE_LOW>;
		high-precision;
	};
```

Signed-off-by: Tomáš Juřena <jurenatomas@gmail.com>

bbilas commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -185,13 +193,44 @@ static int ina230_calibrate(const struct device *dev)
 		return -ENODEV;
 	}
 
+	data->bus_voltage_scaling = INA236_BUS_VOLTAGE_UV_LSB;
```
This can be set at compile time.

darkmoon32 replied:
Moved it there

bbilas commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -185,13 +193,44 @@ static int ina230_calibrate(const struct device *dev)
+
+	goto setup_dev;
+
+setup_dev_ina230:
```
Prepare separate init functions based on the variant. You can follow the i.e https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led/lp50xx.c approach.

darkmoon32 replied:
I was looking for this before I decided to go the "runtime" way. I hope I used the macro correctly in the updated version.

bbilas commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -185,13 +193,44 @@ static int ina230_calibrate(const struct device *dev)
+	data->bus_voltage_scaling = INA236_BUS_VOLTAGE_UV_LSB;
+	data->power_scaling = INA236_POWER_SCALING;
+
+	ret = ina23x_reg_read_16(&config->bus, INA236_REG_DEVICE_ID, &word);
```
It should be called only for INA236.

darkmoon32 replied:
Removed this runtime device check

bbilas commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -249,14 +288,14 @@ static const struct sensor_driver_api ina230_driver_api = {
 
 #define INA230_DRIVER_INIT(inst)                                                                   \
-	static struct ina230_data drv_data_##inst;                                                 \
+	static struct ina230_data drv_data_##inst = {.high_precision =                             \
```
This can be located in the config structure.

darkmoon32 replied:
Moved it there

jurenat commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -20,21 +18,53 @@ LOG_MODULE_REGISTER(INA230, CONFIG_SENSOR_LOG_LEVEL);
+	[ina236] = {
+			.bus_voltage_uv_lsb = INA236_BUS_VOLTAGE_UV_LSB,
+			.power_scaling = INA236_POWER_SCALING,
+			.have_high_precision = true
```
Trailing commas would format it better, in my opinion, but it is not compliant with the code style, and the pre-push hook will reject it with this error `ERROR:OPEN_BRACE: that open brace { should be on the previous line`.

There are these options now:

1. Leave it as it is
2. Add a trailing comma after the last element of the `ina23x_config` array
3. Add a trailing comma into each initialization of an element in the array
4. Add trailing commas after the last element and into each element initialization (will be rejected)
5. Create a macro to initialize the structure
6. Do not use `ina23x_config` at all and pass the configuration in the `DT_INST_FOREACH_STATUS_OKAY_VARGS`

If the second suggestion is applied it will look like this (looks good and compliant):

```c
const static struct ina23x_config ina23x_config[] = {
        [ina230] = {.bus_voltage_uv_lsb = INA230_BUS_VOLTAGE_UV_LSB,
                    .power_scaling = INA230_POWER_SCALING,
                    .have_high_precision = false},
        [ina236] = {.bus_voltage_uv_lsb = INA236_BUS_VOLTAGE_UV_LSB,
                    .power_scaling = INA236_POWER_SCALING,
                    .have_high_precision = true},
};
```

If the third option is applied (elements are misaligned):

```c
const static struct ina23x_config ina23x_config[] = {
        [ina230] =
                { 
                        .bus_voltage_uv_lsb = INA230_BUS_VOLTAGE_UV_LSB,
                        .power_scaling = INA230_POWER_SCALING,
                        .have_high_precision = false,
                },
        [ina236] = {
                .bus_voltage_uv_lsb = INA236_BUS_VOLTAGE_UV_LSB,
                .power_scaling = INA236_POWER_SCALING,
                .have_high_precision = true,
        }};
```

The macro version could be something like this:

```c
#define CREATE_CONFIG(LSB, SCALING, HIGH_PRECISION)                                                \
        {                                                                                          \
                .bus_voltage_uv_lsb = LSB, .power_scaling = SCALING,                               \
                .have_high_precision = HIGH_PRECISION,                                             \
        }

const static struct ina23x_config ina23x_config[] = {
        [ina230] = CREATE_CONFIG(INA230_BUS_VOLTAGE_UV_LSB, INA230_POWER_SCALING, false),
        [ina236] = CREATE_CONFIG(INA236_BUS_VOLTAGE_UV_LSB, INA236_POWER_SCALING, true),
};
```

If you can think of some other way, please suggest it.

bbilas replied:
I like the idea of a dedicated macro. If a new variant were to be added, it would likely only add one extra line into `ina23x_config`.

jurenat replied:
I updated it with the macro version

bbilas commented:
Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.

bbilas commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -20,21 +18,53 @@ LOG_MODULE_REGISTER(INA230, CONFIG_SENSOR_LOG_LEVEL);
+#define INA236_CONFIG_ADC_RANGE BIT(12)
+
+enum ina23x_ids {
+	ina230,
```
Use upper-case characters.

jurenat replied:
Fixed

bbilas commented on dts/bindings/sensor/ti,ina236.yaml:
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright 2021 The Chromium OS Authors
+# Copyright 2021 Grinn
```
This appears to be copy&paste from ina237.

bbilas commented on dts/bindings/sensor/ti,ina236.yaml:
```diff
@@ -0,0 +1,16 @@
+#
+
+description: |
+    TI INA230, INA231 and INA236 Bidirectional Current and Power Monitor.
```
This binding is only for `INA236`.

jurenat replied:
Removed mentions of INA230 and INA231

jurenat commented:
> Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.

First of all, thanks for review.

About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?

bbilas commented:
> > Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.
> 
> First of all, thanks for review.
> 
> About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?
> 
> p.s.: any preferences about [#77403 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/77403#discussion_r1727552908)?

Yes, please add a node into `native_sim.overlay` and extend the ina230 emul device and test to support `INA236` as well.

bbilas commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -20,21 +18,53 @@ LOG_MODULE_REGISTER(INA230, CONFIG_SENSOR_LOG_LEVEL);
+	bool have_high_precision;
+};
+
+const static struct ina23x_config ina23x_config[] = {
```
```suggestion
static const struct ina23x_config ina23x_config[] = {
```

jurenat replied:
Suggestion applied

jurenat committed:
test: sensor: ina230: Add INA236

Update tests to validate INA236 chip

Signed-off-by: Tomáš Juřena <jurenatomas@gmail.com>

jurenat commented:
> > > Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.
> > 
> > 
> > First of all, thanks for review.
> > About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?
> > p.s.: any preferences about [#77403 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/77403#discussion_r1727552908)?
> 
> Yes, please add a node into `native_sim.overlay` and extend the ina230 emul device and test to support `INA236` as well.

I have added the tests for INA236. Please look at it because I don't know if the solution is optimal.

bbilas commented on tests/drivers/sensor/ina230/src/ina230_emul.c:
```diff
@@ -151,22 +158,33 @@ static const struct i2c_emul_api ina230_emul_api_i2c = {
+				{INA230_REG_MASK, 2, 0},                                           \
+				{INA230_REG_ALERT, 2, 0},                                          \
+			},                                                                         \
+		COND_CODE_1(is_236,                                                                \
```
Where does `is_236` come from?

jurenat replied:
Sorry about that, I was trying to use those registers only for INA236 for a moment.

bbilas commented:
> > > > Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.
> > > 
> > > 
> > > First of all, thanks for review.
> > > About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?
> > > p.s.: any preferences about [#77403 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/77403#discussion_r1727552908)?
> > 
> > 
> > Yes, please add a node into `native_sim.overlay` and extend the ina230 emul device and test to support `INA236` as well.
> 
> I have added the tests for INA236. Please look at it because I don't know if the solution is optimal.

I'm only worried about the `get_register` function, where we don't distinguish the INA23X variant. In the case of the same reg addr, we can return the wrong value, but as I can see, this does not happen right now.

jurenat commented:
> > > > > Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.
> > > > 
> > > > 
> > > > First of all, thanks for review.
> > > > About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?
> > > > p.s.: any preferences about [#77403 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/77403#discussion_r1727552908)?
> > > 
> > > 
> > > Yes, please add a node into `native_sim.overlay` and extend the ina230 emul device and test to support `INA236` as well.
> > 
> > 
> > I have added the tests for INA236. Please look at it because I don't know if the solution is optimal.
> 
> I'm only worried about the `get_register` function, where we don't distinguish the INA23X variant. In the case of the same reg addr, we can return the wrong value, but as I can see, this does not happen right now.

If that concerns you, I think it could be added in `ina230_emul_data` per device type as a pointer.

bbilas commented:
> > > > > > Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.
> > > > > 
> > > > > 
> > > > > First of all, thanks for review.
> > > > > About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?
> > > > > p.s.: any preferences about [#77403 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/77403#discussion_r1727552908)?
> > > > 
> > > > 
> > > > Yes, please add a node into `native_sim.overlay` and extend the ina230 emul device and test to support `INA236` as well.
> > > 
> > > 
> > > I have added the tests for INA236. Please look at it because I don't know if the solution is optimal.
> > 
> > 
> > I'm only worried about the `get_register` function, where we don't distinguish the INA23X variant. In the case of the same reg addr, we can return the wrong value, but as I can see, this does not happen right now.
> 
> If that concerns you, I think it could be added in `ina230_emul_data` per device type as a pointer.

I was thinking about this solution. Could you add it?

PS It seems you need to rebase with main.

jurenat commented:
> > > > > > > Add a new node with `ina236` into `tests/drivers/sensor/ina230` to be built by CI.
> > > > > > 
> > > > > > 
> > > > > > First of all, thanks for review.
> > > > > > About your comment, I am not sure what you mean. Should I add a node in the DTS, update the tests and write the emul device?
> > > > > > p.s.: any preferences about [#77403 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/77403#discussion_r1727552908)?
> > > > > 
> > > > > 
> > > > > Yes, please add a node into `native_sim.overlay` and extend the ina230 emul device and test to support `INA236` as well.
> > > > 
> > > > 
> > > > I have added the tests for INA236. Please look at it because I don't know if the solution is optimal.
> > > 
> > > 
> > > I'm only worried about the `get_register` function, where we don't distinguish the INA23X variant. In the case of the same reg addr, we can return the wrong value, but as I can see, this does not happen right now.
> > 
> > 
> > If that concerns you, I think it could be added in `ina230_emul_data` per device type as a pointer.
> 
> I was thinking about this solution. Could you add it?
> 
> PS It seems you need to rebase with main.

Rebased and added `get_register` callback. Please look at it once more.

bbilas commented on tests/drivers/sensor/ina230/src/ina230_emul.c:
```diff
@@ -30,23 +29,38 @@ struct ina230_emul_cfg {
 };
 
-static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
+static struct ina230_reg *ina230_get_register(struct ina230_emul_data *data, int reg)
```
I still don't see how it should work if someone requests the reg address from INA230 using the INA236 variant. `ina230_get_register` is still common for both variants as well as `ina230_regs` and `ina236_regs`. IMO. we should define INA23X_regs array only for specified variants and check what is what. I see this something like that:

<pre>
static struct ina230_reg *ina230_get_register(struct ina230_emul_data *data, int reg)
{
	for (int i = 0; i < data->regs_len; i++) {
		if (data->regs[i].id == reg) {
			return &data->regs[i];
		}
	}

	return NULL;
}
</pre>

I would add a `regs` pointer pointing to the regs array for a specific variant. This requires distinguishing between variants in the `INA230_EMUL` macro to correctly assign the regs array (you did something similar before with `is_ina236` variable).

jurenat replied:
> I still don't see how it should work if someone requests the reg address from INA230 using the INA236 variant. `ina230_get_register` is still common for both variants as well as `ina230_regs` and `ina236_regs`. IMO. we should define INA23X_regs array only for specified variants and check what is what. I see this something like that:
> 
> static struct ina230_reg *ina230_get_register(struct ina230_emul_data *data, int reg)
> {
> 	for (int i = 0; i < data->regs_len; i++) {
> 		if (data->regs[i].id == reg) {
> 			return &data->regs[i];
> 		}
> 	}
> 
> 	return NULL;
> }
> I would add a `regs` pointer pointing to the regs array for a specific variant. This requires distinguishing between variants in the `INA230_EMUL` macro to correctly assign the regs array (you did something similar before with `is_ina236` variable).

Oh, I  misunderstood you before, you are correct about it. We can create `base_regs` and `extra_regs`, where `base_regs` will store the original INA230 registers and their default values, and the `extra_regs`  will hold the "overloaded" registers with different defaults and possibly some extra registers. This array must be **0xff** bytes long so we can hold all possible addresses. Then the `get_register` function could look like this:

```c
static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
{
        struct ina230_reg * c_reg = data->extra_regs;
        for (size_t i = 0; i <  ARRAY_SIZE(data->extra_regs); i++,c_reg++) {
                // valid entry?
                if (!c_reg->bytes) {
                        break;
                }

                if (c_reg->id == reg) {
                        return c_reg;
                }
        }

        c_reg = data->base_regs;
        for (size_t i = 0; i <  INA230_REGISTER_COUNT; i++,c_reg++) {
                if (c_reg->id == reg) {
                        return c_reg;
                }
        }

        return NULL;
}
```

The initialization can look like this:

```c
#define CREATE_INA236_REGS                                                                         \
        {                                                                                          \
                {INA236_REG_MANUFACTURER_ID, 2, 0x449}, {INA236_REG_DEVICE_ID, 2, 0xa080},         \
                        {INA230_REG_MASK, 2, 0x10},                                                \
        }

#define INA230_EMUL(n, v)                                                                          \
        static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
                .addr = DT_INST_REG_ADDR(n),                                                       \
        };                                                                                         \
        static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
                .ina230_regs =                                                                     \
                        {                                                                          \
                                {INA230_REG_CONFIG, 2, 0x4127},                                    \
                                {INA230_REG_SHUNT_VOLT, 2, 0},                                     \
                                {INA230_REG_BUS_VOLT, 2, 0},                                       \
                                {INA230_REG_POWER, 2, 0},                                          \
                                {INA230_REG_CURRENT, 2, 0},                                        \
                                {INA230_REG_CALIB, 2, 0},                                          \
                                {INA230_REG_MASK, 2, 0},                                           \
                                {INA230_REG_ALERT, 2, 0},                                          \
                        },                                                                         \
                .ina236_regs = COND_CODE_0(v, ({}), (CREATE_INA23##v##_REGS)),                     \
                .get_register = ina23##v##_get_register,                                           \
        };                                                                                         \
        EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
                            &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)
```

I am just not sure about the `COND_CODE_0` part where it is not a flag that is being checked. What do you think about this solution?

bbilas replied:
Let's try to simplify it ;) Could you test the following pseudo-code?

<pre>
struct ina230_emul_data {
	struct ina23x_config *ina23x_config;
	const size_t ina23x_config_len;
};

static struct ina23x_config ina230_config[] = {
	{INA230_REG_CONFIG, 2, 0x4127},                                    
	{INA230_REG_SHUNT_VOLT, 2, 0},                                     
	{INA230_REG_BUS_VOLT, 2, 0},                                       
	{INA230_REG_POWER, 2, 0},                                          
	{INA230_REG_CURRENT, 2, 0},                                        
	{INA230_REG_CALIB, 2, 0},                                          
	{INA230_REG_MASK, 2, 0},                                           
	{INA230_REG_ALERT, 2, 0},                                     
};		

static struct ina23x_config ina236_config[] = {
	{INA236_REG_MANUFACTURER_ID, 2, 0x449},
	{INA236_REG_DEVICE_ID, 2, 0xa080},
};		
          			
#define INA230_EMUL(n, id, v)                                                                          \
	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
		.addr = DT_INST_REG_ADDR(n),                                                       \
	};                                                                                         \
	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
		.ina23x_config = &type##_config; \	
		.ina23x_config_len = ARRAY_SIZE(type##_config); \	
	};                                                                                         \
	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)

#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_ina230
DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina230, 0)

#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_ina236
DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina236, 6)


static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
{
	for (int i = 0; i < data->ina23x_config_len; i++) {
		if (data->ina23x_config[i]->id == reg) {
			return &data->ina23x_config[i];
		}
	}

	return NULL;
}
<pre>

jurenat replied:
> Let's try to simplify it ;) Could you test the following pseudo-code?
> 
> struct ina230_emul_data {
> 	struct ina23x_config *ina23x_config;
> 	const size_t ina23x_config_len;
> };
> 
> static struct ina23x_config ina230_config[] = {
> 	{INA230_REG_CONFIG, 2, 0x4127},                                    
> 	{INA230_REG_SHUNT_VOLT, 2, 0},                                     
> 	{INA230_REG_BUS_VOLT, 2, 0},                                       
> 	{INA230_REG_POWER, 2, 0},                                          
> 	{INA230_REG_CURRENT, 2, 0},                                        
> 	{INA230_REG_CALIB, 2, 0},                                          
> 	{INA230_REG_MASK, 2, 0},                                           
> 	{INA230_REG_ALERT, 2, 0},                                     
> };		
> 
> static struct ina23x_config ina236_config[] = {
> 	{INA236_REG_MANUFACTURER_ID, 2, 0x449},
> 	{INA236_REG_DEVICE_ID, 2, 0xa080},
> };		
>           			
> #define INA230_EMUL(n, id, v)                                                                          \
> 	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
> 		.addr = DT_INST_REG_ADDR(n),                                                       \
> 	};                                                                                         \
> 	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
> 		.ina23x_config = &type##_config; \	
> 		.ina23x_config_len = ARRAY_SIZE(type##_config); \	
> 	};                                                                                         \
> 	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
> 			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)
> 
> #undef DT_DRV_COMPAT
> #define DT_DRV_COMPAT ti_ina230
> DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina230, 0)
> 
> #undef DT_DRV_COMPAT
> #define DT_DRV_COMPAT ti_ina236
> DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina236, 6)
> 
> 
> static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
> {
> 	for (int i = 0; i < data->ina23x_config_len; i++) {
> 		if (data->ina23x_config[i]->id == reg) {
> 			return &data->ina23x_config[i];
> 		}
> 	}
> 
> 	return NULL;
> }

If I integrated your code correctly then this is not working for INA236.

How can this work with registers for INA230 when 236 is used? This code only takes a pointer for `ina236_config` registers: 

```c
static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
        .ina23x_config = &type##_config; \	
        .ina23x_config_len = ARRAY_SIZE(type##_config); \	
};
```

P.S.: is it OK to share the register values (we take the pointer to a static array) between instances of the same type of emulated devices?

bbilas replied:
It takes the pointer to the specified reg array based on the id. Why would you want to use registers from INA230 when INA236 is being used?

jurenat replied:
Because the tests are using them for mocks and to have a complete set of registers defined by chip. How else can we test the different resolutions of the chips?

What exactly are we trying to achieve? I thought that I needed to create a complete virtual INA236 device with all its registers so we could mock them in tests if needed.

bbilas replied:
Oh, I missed that completely. In that case, we can use your previous approach or extend the `ina236_config` about those registers. I'm sorry for the confusion.

jurenat replied:
It would, I just did not want to mix that up, because it would not fit the array (`INA230_REGISTER_COUNT` is 8). Also, I think `get_register` should return `NULL` when a register not provided by the device is requested.

jurenat replied:
Are the registers split into two arrays blocking the pull? Can I do something to have this approved?

bbilas replied:
Some of the checks failed, try to rebase with the latest main and force push. Regarding the implementation, I'm still thinking about the pre-pre version with 2 separate for loops (the current one seems too complicated to me though).

jurenat replied:
Ok, what if we solve it like this:

```c
struct ina230_emul_data {
	struct ina230_reg *regs;
};

static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
{
	struct ina230_reg *c_reg = data->regs;

	while (c_reg->bytes) {
		if (c_reg->id == reg) {
			return c_reg;
		}
		c_reg++;
	}

	return NULL;
}

#define CREATE_INA230_REGS                                                                         \
	{INA230_REG_CONFIG, 2, 0x4127}, {INA230_REG_SHUNT_VOLT, 2, 0},                             \
		{INA230_REG_BUS_VOLT, 2, 0}, {INA230_REG_POWER, 2, 0}, {INA230_REG_CURRENT, 2, 0}, \
		{INA230_REG_CALIB, 2, 0}, {INA230_REG_MASK, 2, 0}, {INA230_REG_ALERT, 2, 0},

#define CREATE_INA236_REGS                                                                         \
	CREATE_INA230_REGS, {INA236_REG_MANUFACTURER_ID, 2, 0x449},                                \
	{                                                                                          \
		INA236_REG_DEVICE_ID, 2, 0xa080                                                    \
	}

#define INA230_EMUL(n, v)                                                                          \
	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
		.addr = DT_INST_REG_ADDR(n),                                                       \
	};                                                                                         \
	static struct ina230_reg ina23##v##_regs_##n[] = {CREATE_INA23##v##_REGS, {}};             \
	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
		.regs = (struct ina230_reg *)ina23##v##_regs_##n};                                 \
	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)
```

With this solution there is only one array and one loop to lookup the register in the mocked device. There is a little readability issue in `CREATE_INA23x_REGS` macros, but this can be hopefully handled better later.

bbilas replied:
Yea, readability is pretty bad, but the single array implementation looks much better now. 

btw. did you use clang formatter or is it done by hand?

jurenat replied:
Unfortunately, it is the output from clang-format-20.

bbilas replied:
You can try to improve that by hand as clang formatting is not a must-have.

jurenat replied:
Is this better?

```c
/* clang-format off */

#define CREATE_INA230_REGS                                                                         \
	{INA230_REG_CONFIG, 2, 0x4127},                                                            \
	{INA230_REG_SHUNT_VOLT, 2, 0},                                                             \
	{INA230_REG_BUS_VOLT, 2, 0},                                                               \
	{INA230_REG_POWER, 2, 0},                                                                  \
	{INA230_REG_CURRENT, 2, 0},                                                                \
	{INA230_REG_CALIB, 2, 0},                                                                  \
	{INA230_REG_MASK, 2, 0},                                                                   \
	{INA230_REG_ALERT, 2, 0}

#define CREATE_INA236_REGS                                                                         \
	CREATE_INA230_REGS,                                                                        \
	{INA236_REG_MANUFACTURER_ID, 2, 0x449},                                                    \
	{INA236_REG_DEVICE_ID, 2, 0xa080}                                                          \

#define INA230_EMUL(n, v)                                                                          \
	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
		.addr = DT_INST_REG_ADDR(n),                                                       \
	};                                                                                         \
	static struct ina230_reg ina23##v##_regs_##n[] = {CREATE_INA23##v##_REGS, {}};             \
	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
		.regs = (struct ina230_reg *)ina23##v##_regs_##n};                                 \
	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)

/* clang-format on */
```

bbilas replied:
Looks pretty good now!

jeppenodgaard commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -168,13 +201,26 @@ static int ina230_attr_get(const struct device *dev, enum sensor_channel chan,
+
 	/* See datasheet "Programming" section */
-	ret = ina23x_reg_write(&config->bus, INA230_REG_CALIB, config->cal);
+	ret = ina23x_reg_write(&config->bus, INA230_REG_CALIB, cal);
```
This seems like an unnecessary change for this commit. I think it should either be reverted or have it's own commit with a change reason.

jurenat replied:
The reason is the high-precision option that can be enabled for INA236 and it requires to divide the `SHUNT_CAL` by 4. Do you have a better idea of how to do this?

jeppenodgaard replied:
I mean adding a variable to hold the value of `config->cal`. There's no functional change here, just coding style.

jurenat replied:
I think I do not understand you, but I reverted the `cal` usage and used `config->cal` instead. The division of the LSB is done when creating an instance of the device.

jeppenodgaard commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -185,14 +231,19 @@ static int ina230_calibrate(const struct device *dev)
+		config_raw |= INA236_CONFIG_ADC_RANGE;
+	}
+
+	ret = ina23x_reg_write(&config->bus, INA230_REG_CONFIG, config_raw);
```
This seems like an unnecessary change for this commit. I think it should either be reverted or have it's own commit with a change reason.

jurenat replied:
Do you mean that the support for high-precision should be added in a separate commit?

jeppenodgaard replied:
I mean adding a variable to hold the value of config->config. There's no functional change here, just coding style.

jurenat replied:
I think I do not understand you, but I reverted the `config_raw` usage and used `config->config` instead. Enabling the `high-precision` mode is done when creating an instance of the device.

jeppenodgaard commented on dts/bindings/sensor/ti,ina230.yaml:
```diff
@@ -80,3 +80,8 @@ properties:
     # default alert limit is 0V
     description: Alert register, default matches the power-on reset value
+
+  high-precision:
```
Looks like this belongs in `ina236.yaml` since that is the device that supports it.

jurenat replied:
Good catch, I will move this into ina236.yaml.

jeppenodgaard commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -20,21 +18,56 @@ LOG_MODULE_REGISTER(INA230, CONFIG_SENSOR_LOG_LEVEL);
+struct ina23x_config {
+	uint16_t bus_voltage_uv_lsb;
+	uint8_t power_scaling;
+	bool have_high_precision;
```
I think this can be removed, if it is possible to set via a property in the dts. If it cannot be set then the device does not support high precision mode.

jeppenodgaard reviewed (commented):
Nice with the added ina236 functionality.  I got a few comments but overall I think it looks good.

jurenat commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -20,21 +18,56 @@ LOG_MODULE_REGISTER(INA230, CONFIG_SENSOR_LOG_LEVEL);
+	bool have_high_precision;
+};
+
+static const struct ina23x_config ina23x_config[] = {
```
There can be multiple instances of any combination of INA230 and INA236. Because of this, we need to know what configuration we need to apply so this array holds the configuration per device type and it is shared between instances of the same device type to save some memory.

jeppenodgaard replied:
Without testing I guess it would require at least 3 devices to break even or save memory. Below 3 should cost more memory since you need the additional `void *`. The `void *` also increases code complexity IMO.

jurenat replied:
Removed.

jurenat commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -185,14 +231,19 @@ static int ina230_calibrate(const struct device *dev)
 
-	ret = ina23x_reg_write(&config->bus, INA230_REG_CONFIG, config->config);
+	if (ina230_is_high_precision(config)) {
+		config_raw |= INA236_CONFIG_ADC_RANGE;
```
The `INA236_CONFIG_ADC_RANGE` defines only INA236. I think it would be better to rename the `ina230_is_high_precision` function to `ina23x_is_high_precision` instead.

jeppenodgaard reviewed (changes requested):
One last thing. Otherwise LGTM.

jurenat commented on drivers/sensor/ti/ina23x/ina230.c:
```diff
@@ -248,23 +248,32 @@ static const struct sensor_driver_api ina230_driver_api = {
+			  (DT_INST_ENUM_IDX(inst, vbus_conversion_time_us) << 6) |                 \
+			  (DT_INST_ENUM_IDX(inst, vshunt_conversion_time_us) << 3) |               \
+			  DT_INST_ENUM_IDX(inst, adc_mode) |                                       \
+			  (DT_INST_PROP_OR(inst, high_precision, 0) << 12),                        \
```
Are you sure `_OR` is not required? When I remove it I get a compile error `error: ‘DT_N_S_i2c_100_S_ina230_40_P_high_precision’ undeclared here (not in a function); did you mean ‘DT_N_S_i2c_100_S_ina236_41_P_high_precision’?`

jurenat replied:
@jeppenodgaard I will move the line as requested, but I do not think I can replace `DT_INST_PROP_OR` with `DT_INST_PROP`. Could you let me know if you are sure about it?

MaureenHelm replied:
`DT_INST_PROP` should be sufficient for a boolean, but if you get a compile error then it's fine to use `DT_INST_PROP_OR`

jurenat replied:
Ok, I rebased the code and moved the line.

`DT_INST_PROP` should be enough if the property is defined in DTS binding, but it is not determined from INA230.
